rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/HostsFileBasedInfrastructureManager.java

```diff
@@ -384,211 +382,180 @@ private int getRetriesLeft(int retries) {
+     * The node name involves the host that was written in the host file, and
+     * that si used as a identifying key to track the node afterwards.
      */
-    private InetAddress removeNodeAndReturnHost(final String nodeName, final String nodeUrl, final Node node) {
-        return setPersistedInfraVariable(new PersistedInfraVariablesHandler<InetAddress>() {
-            @Override
-            public InetAddress handle() {
-                InetAddress host = getRegisteredNodes().remove(nodeName);
-                if (host != null) {
-                    logger.debug("Removing node " + nodeUrl + " from " + this.getClass().getSimpleName());
-                    addRemovedHost(host);
-                    if (!getRegisteredNodes().containsValue(host) && node != null) {
-                        try {
-                            killNodeImpl(node, host);
-                        } catch (Exception e) {
-                            logger.trace("An exception occurred during node kill", e);
-                        }
-                    }
-                }
-                return host;
-            }
-        });
-    }
+    protected class NodeNameBuilder implements Serializable {
+
+        private static final String SSH_NODE_NAME_PREFIX = "SSH";
+
+        private static final String NODE_NAME_BUILDING_BLOCK_SEPARATOR = "-";
+
+        private static final int HOST_BUILDING_BLOCK_POSITION = 2;
+
+        private static final char HOST_ADDRESS_COMPONENT_SEPARATOR = '.';
 
-    private void addRemovedHost(InetAddress host) {
-        Integer retrieved = getRemovedHosts().get(host);
-        if (retrieved == null) {
-            retrieved = 0;
+        private static final char CONVERTED_HOST_ADDRESS_COMPONENT_SEPARATOR = '_';
+
+        protected String generateNodeName(HostTracker hostTracker) {
+            String hostInFile = hostTracker.getHostInFile();
+            String compliantNodeName = convertHostInFileIntoCompliantNodeName(hostInFile);
+            return SSH_NODE_NAME_PREFIX + NODE_NAME_BUILDING_BLOCK_SEPARATOR + nodeSource.getName() +
+                   NODE_NAME_BUILDING_BLOCK_SEPARATOR + compliantNodeName + NODE_NAME_BUILDING_BLOCK_SEPARATOR +
+                   ProActiveCounter.getUniqID();
         }
-        getRemovedHosts().put(host, ++retrieved);
-    }
 
-    private int computeExpectedNbNodesForHost(InetAddress host) {
-        Collection<InetAddress> address = getHostAddressPerNodeUrl().values();
-        int nbNodesForHost = 0;
-        for (InetAddress inetAddress : address) {
-            if (inetAddress.equals(host)) {
-                nbNodesForHost++;
-            }
+        protected String extractHostFromNodeName(String nodeName) {
+            String[] nodeNameBuildingBlocks = nodeName.split(NODE_NAME_BUILDING_BLOCK_SEPARATOR);
+            String compliantNodeName = nodeNameBuildingBlocks[HOST_BUILDING_BLOCK_POSITION];
+            return convertCompliantNodeNameIntoHostInFile(compliantNodeName);
+        }
```

Comment:
I prefer when internal classes are defined at the end of the class, for readability

Reply:
applied